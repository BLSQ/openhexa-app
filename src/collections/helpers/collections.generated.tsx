import * as Types from '../../graphql-types';

import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
const defaultOptions = {} as const;
export type AddDhis2DataElementToCollectionMutationVariables = Types.Exact<{
  input: Types.AddDhis2DataElementToCollectionInput;
}>;


export type AddDhis2DataElementToCollectionMutation = { __typename?: 'Mutation', addToCollection: { __typename?: 'AddDHIS2DataElementToCollectionResult', success: boolean, errors: Array<Types.AddDhis2DataElementToCollectionError> } };

export type AddS3ObjectToCollectionMutationVariables = Types.Exact<{
  input: Types.AddS3ObjectToCollectionInput;
}>;


export type AddS3ObjectToCollectionMutation = { __typename?: 'Mutation', addToCollection: { __typename?: 'AddS3ObjectToCollectionResult', success: boolean, errors: Array<Types.AddS3ObjectToCollectionError> } };

export type RemoveDhis2DataElementFromCollectionMutationVariables = Types.Exact<{
  input: Types.RemoveDhis2DataElementFromCollectionInput;
}>;


export type RemoveDhis2DataElementFromCollectionMutation = { __typename?: 'Mutation', removeFromCollection: { __typename?: 'RemoveDHIS2DataElementFromCollectionResult', success: boolean, errors: Array<Types.RemoveDhis2DataElementFromCollectionError> } };

export type RemoveS3ObjectFromCollectionMutationVariables = Types.Exact<{
  input: Types.RemoveS3ObjectFromCollectionInput;
}>;


export type RemoveS3ObjectFromCollectionMutation = { __typename?: 'Mutation', removeFromCollection: { __typename?: 'RemoveS3ObjectFromCollectionResult', success: boolean, errors: Array<Types.RemoveS3ObjectFromCollectionError> } };

export type CreateCollectionMutationVariables = Types.Exact<{
  input: Types.CreateCollectionInput;
}>;


export type CreateCollectionMutation = { __typename?: 'Mutation', createCollection: { __typename?: 'CreateCollectionResult', success: boolean, errors: Array<Types.CreateCollectionError>, collection?: { __typename?: 'Collection', id: string, name: string } | null } };


export const AddDhis2DataElementToCollectionDocument = gql`
    mutation addDHIS2DataElementToCollection($input: AddDHIS2DataElementToCollectionInput!) {
  addToCollection: addDHIS2DataElementToCollection(input: $input) {
    success
    errors
  }
}
    `;
export type AddDhis2DataElementToCollectionMutationFn = Apollo.MutationFunction<AddDhis2DataElementToCollectionMutation, AddDhis2DataElementToCollectionMutationVariables>;

/**
 * __useAddDhis2DataElementToCollectionMutation__
 *
 * To run a mutation, you first call `useAddDhis2DataElementToCollectionMutation` within a React component and pass it any options that fit your needs.
 * When your component renders, `useAddDhis2DataElementToCollectionMutation` returns a tuple that includes:
 * - A mutate function that you can call at any time to execute the mutation
 * - An object with fields that represent the current status of the mutation's execution
 *
 * @param baseOptions options that will be passed into the mutation, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options-2;
 *
 * @example
 * const [addDhis2DataElementToCollectionMutation, { data, loading, error }] = useAddDhis2DataElementToCollectionMutation({
 *   variables: {
 *      input: // value for 'input'
 *   },
 * });
 */
export function useAddDhis2DataElementToCollectionMutation(baseOptions?: Apollo.MutationHookOptions<AddDhis2DataElementToCollectionMutation, AddDhis2DataElementToCollectionMutationVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useMutation<AddDhis2DataElementToCollectionMutation, AddDhis2DataElementToCollectionMutationVariables>(AddDhis2DataElementToCollectionDocument, options);
      }
export type AddDhis2DataElementToCollectionMutationHookResult = ReturnType<typeof useAddDhis2DataElementToCollectionMutation>;
export type AddDhis2DataElementToCollectionMutationResult = Apollo.MutationResult<AddDhis2DataElementToCollectionMutation>;
export type AddDhis2DataElementToCollectionMutationOptions = Apollo.BaseMutationOptions<AddDhis2DataElementToCollectionMutation, AddDhis2DataElementToCollectionMutationVariables>;
export const AddS3ObjectToCollectionDocument = gql`
    mutation addS3ObjectToCollection($input: AddS3ObjectToCollectionInput!) {
  addToCollection: addS3ObjectToCollection(input: $input) {
    success
    errors
  }
}
    `;
export type AddS3ObjectToCollectionMutationFn = Apollo.MutationFunction<AddS3ObjectToCollectionMutation, AddS3ObjectToCollectionMutationVariables>;

/**
 * __useAddS3ObjectToCollectionMutation__
 *
 * To run a mutation, you first call `useAddS3ObjectToCollectionMutation` within a React component and pass it any options that fit your needs.
 * When your component renders, `useAddS3ObjectToCollectionMutation` returns a tuple that includes:
 * - A mutate function that you can call at any time to execute the mutation
 * - An object with fields that represent the current status of the mutation's execution
 *
 * @param baseOptions options that will be passed into the mutation, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options-2;
 *
 * @example
 * const [addS3ObjectToCollectionMutation, { data, loading, error }] = useAddS3ObjectToCollectionMutation({
 *   variables: {
 *      input: // value for 'input'
 *   },
 * });
 */
export function useAddS3ObjectToCollectionMutation(baseOptions?: Apollo.MutationHookOptions<AddS3ObjectToCollectionMutation, AddS3ObjectToCollectionMutationVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useMutation<AddS3ObjectToCollectionMutation, AddS3ObjectToCollectionMutationVariables>(AddS3ObjectToCollectionDocument, options);
      }
export type AddS3ObjectToCollectionMutationHookResult = ReturnType<typeof useAddS3ObjectToCollectionMutation>;
export type AddS3ObjectToCollectionMutationResult = Apollo.MutationResult<AddS3ObjectToCollectionMutation>;
export type AddS3ObjectToCollectionMutationOptions = Apollo.BaseMutationOptions<AddS3ObjectToCollectionMutation, AddS3ObjectToCollectionMutationVariables>;
export const RemoveDhis2DataElementFromCollectionDocument = gql`
    mutation removeDHIS2DataElementFromCollection($input: RemoveDHIS2DataElementFromCollectionInput!) {
  removeFromCollection: removeDHIS2DataElementFromCollection(input: $input) {
    success
    errors
  }
}
    `;
export type RemoveDhis2DataElementFromCollectionMutationFn = Apollo.MutationFunction<RemoveDhis2DataElementFromCollectionMutation, RemoveDhis2DataElementFromCollectionMutationVariables>;

/**
 * __useRemoveDhis2DataElementFromCollectionMutation__
 *
 * To run a mutation, you first call `useRemoveDhis2DataElementFromCollectionMutation` within a React component and pass it any options that fit your needs.
 * When your component renders, `useRemoveDhis2DataElementFromCollectionMutation` returns a tuple that includes:
 * - A mutate function that you can call at any time to execute the mutation
 * - An object with fields that represent the current status of the mutation's execution
 *
 * @param baseOptions options that will be passed into the mutation, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options-2;
 *
 * @example
 * const [removeDhis2DataElementFromCollectionMutation, { data, loading, error }] = useRemoveDhis2DataElementFromCollectionMutation({
 *   variables: {
 *      input: // value for 'input'
 *   },
 * });
 */
export function useRemoveDhis2DataElementFromCollectionMutation(baseOptions?: Apollo.MutationHookOptions<RemoveDhis2DataElementFromCollectionMutation, RemoveDhis2DataElementFromCollectionMutationVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useMutation<RemoveDhis2DataElementFromCollectionMutation, RemoveDhis2DataElementFromCollectionMutationVariables>(RemoveDhis2DataElementFromCollectionDocument, options);
      }
export type RemoveDhis2DataElementFromCollectionMutationHookResult = ReturnType<typeof useRemoveDhis2DataElementFromCollectionMutation>;
export type RemoveDhis2DataElementFromCollectionMutationResult = Apollo.MutationResult<RemoveDhis2DataElementFromCollectionMutation>;
export type RemoveDhis2DataElementFromCollectionMutationOptions = Apollo.BaseMutationOptions<RemoveDhis2DataElementFromCollectionMutation, RemoveDhis2DataElementFromCollectionMutationVariables>;
export const RemoveS3ObjectFromCollectionDocument = gql`
    mutation removeS3ObjectFromCollection($input: RemoveS3ObjectFromCollectionInput!) {
  removeFromCollection: removeS3ObjectFromCollection(input: $input) {
    success
    errors
  }
}
    `;
export type RemoveS3ObjectFromCollectionMutationFn = Apollo.MutationFunction<RemoveS3ObjectFromCollectionMutation, RemoveS3ObjectFromCollectionMutationVariables>;

/**
 * __useRemoveS3ObjectFromCollectionMutation__
 *
 * To run a mutation, you first call `useRemoveS3ObjectFromCollectionMutation` within a React component and pass it any options that fit your needs.
 * When your component renders, `useRemoveS3ObjectFromCollectionMutation` returns a tuple that includes:
 * - A mutate function that you can call at any time to execute the mutation
 * - An object with fields that represent the current status of the mutation's execution
 *
 * @param baseOptions options that will be passed into the mutation, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options-2;
 *
 * @example
 * const [removeS3ObjectFromCollectionMutation, { data, loading, error }] = useRemoveS3ObjectFromCollectionMutation({
 *   variables: {
 *      input: // value for 'input'
 *   },
 * });
 */
export function useRemoveS3ObjectFromCollectionMutation(baseOptions?: Apollo.MutationHookOptions<RemoveS3ObjectFromCollectionMutation, RemoveS3ObjectFromCollectionMutationVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useMutation<RemoveS3ObjectFromCollectionMutation, RemoveS3ObjectFromCollectionMutationVariables>(RemoveS3ObjectFromCollectionDocument, options);
      }
export type RemoveS3ObjectFromCollectionMutationHookResult = ReturnType<typeof useRemoveS3ObjectFromCollectionMutation>;
export type RemoveS3ObjectFromCollectionMutationResult = Apollo.MutationResult<RemoveS3ObjectFromCollectionMutation>;
export type RemoveS3ObjectFromCollectionMutationOptions = Apollo.BaseMutationOptions<RemoveS3ObjectFromCollectionMutation, RemoveS3ObjectFromCollectionMutationVariables>;
export const CreateCollectionDocument = gql`
    mutation CreateCollection($input: CreateCollectionInput!) {
  createCollection(input: $input) {
    success
    errors
    collection {
      id
      name
    }
  }
}
    `;
export type CreateCollectionMutationFn = Apollo.MutationFunction<CreateCollectionMutation, CreateCollectionMutationVariables>;

/**
 * __useCreateCollectionMutation__
 *
 * To run a mutation, you first call `useCreateCollectionMutation` within a React component and pass it any options that fit your needs.
 * When your component renders, `useCreateCollectionMutation` returns a tuple that includes:
 * - A mutate function that you can call at any time to execute the mutation
 * - An object with fields that represent the current status of the mutation's execution
 *
 * @param baseOptions options that will be passed into the mutation, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options-2;
 *
 * @example
 * const [createCollectionMutation, { data, loading, error }] = useCreateCollectionMutation({
 *   variables: {
 *      input: // value for 'input'
 *   },
 * });
 */
export function useCreateCollectionMutation(baseOptions?: Apollo.MutationHookOptions<CreateCollectionMutation, CreateCollectionMutationVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useMutation<CreateCollectionMutation, CreateCollectionMutationVariables>(CreateCollectionDocument, options);
      }
export type CreateCollectionMutationHookResult = ReturnType<typeof useCreateCollectionMutation>;
export type CreateCollectionMutationResult = Apollo.MutationResult<CreateCollectionMutation>;
export type CreateCollectionMutationOptions = Apollo.BaseMutationOptions<CreateCollectionMutation, CreateCollectionMutationVariables>;