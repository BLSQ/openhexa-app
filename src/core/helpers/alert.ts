export enum AlertType {
  success,
  error,
  info,
  warning,
}

export function displayAlert(message: string, type?: AlertType) {
  if (!window) {
    console.warn("displayAlert cannot be called on the server.");
    return;
  }
  const event = new CustomEvent("displayAlert", {
    detail: {
      message,
      type: type ?? AlertType.info,
    },
  });

  window.dispatchEvent(event);
}
