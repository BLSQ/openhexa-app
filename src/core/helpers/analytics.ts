import { NextRequest } from "next/server";
import getConfig from "next/config";

const { publicRuntimeConfig } = getConfig();

interface TrackEventProperties {
  [key: string]: any;
}

async function sendEvent(
  event: string,
  properties: TrackEventProperties,
  headers?: Headers,
): Promise<void> {
  if (publicRuntimeConfig.DISABLE_ANALYTICS === "true") {
    return;
  }
  const res = await fetch(
    `${publicRuntimeConfig.OPENHEXA_BACKEND_URL ?? ""}/analytics/track/`,
    {
      method: "POST",
      priority: "low",
      headers: headers,
      credentials: "include",
      body: JSON.stringify({ event, properties }),
    },
  );

  if (!res.ok) {
    throw new Error(`Failed to send event: ${res.statusText}`);
  }
}

export async function trackEvent(
  event: string,
  properties: TrackEventProperties,
): Promise<void> {
  if (typeof window === "undefined") {
    // We do not support tracking events on the server
    return;
  }
  try {
    await sendEvent(event, properties);
  } catch (error) {
    console.error("Error tracking event:", error);
  }
}

export async function pageView(request: NextRequest) {
  // This is to be called on the server side only.
  if (typeof window !== "undefined") {
    return;
  }
  try {
    const headers = request.headers;
    headers.set("content-type", "application/json");

    await sendEvent(
      "pageview",
      {
        pathname: request.nextUrl.pathname,
        searchParams: request.nextUrl.searchParams.toString(),
      },
      headers,
    );
  } catch (error) {
    console.error("Error tracking event:", error);
  }
}
