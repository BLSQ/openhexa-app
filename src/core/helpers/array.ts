export function ensureArray<T>(maybeArray: T[] | T): T[] {
  if (typeof maybeArray === "undefined") {
    return [];
  } else if (maybeArray === null) {
    return [];
  } else if (Array.isArray(maybeArray)) {
    return maybeArray;
  } else {
    return new Array(maybeArray);
  }
}
