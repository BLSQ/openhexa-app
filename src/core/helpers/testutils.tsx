import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import { faker } from "@faker-js/faker";
import { ByRoleOptions, screen } from "@testing-library/react";
import { ReactNode } from "react";

type TestAppProps = {
  children: ReactNode;
  mocks?: MockedResponse[];
};

jest.mock("next/router", () => require("next-router-mock"));
// This is needed for mocking 'next/link':
jest.mock("next/dist/client/router", () => require("next-router-mock"));
jest.mock("identity/hooks/useMe");

export function TestApp(props: TestAppProps) {
  return (
    <MockedProvider addTypename={false} mocks={props.mocks ?? []}>
      {props.children}
    </MockedProvider>
  );
}

export const LOGGED_IN_USER = {
  id: faker.datatype.uuid(),
  firstName: faker.name.firstName,
  lastName: faker.name.lastName,
  avatar: {
    initials: "AE",
    color: "red",
  },
};

export async function waitForDialog(options?: ByRoleOptions) {
  const dialog = await screen.getByRole("dialog", options);
  return dialog;
}
