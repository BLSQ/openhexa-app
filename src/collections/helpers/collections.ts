import { gql } from "@apollo/client";
import { getApolloClient } from "core/helpers/apollo";
import { CreateCollectionInput } from "graphql-types";
import {
  CreateCollectionMutationVariables,
  CreateCollectionMutation,
  AddDhis2DataElementToCollectionMutation,
  AddDhis2DataElementToCollectionMutationVariables,
  AddS3ObjectToCollectionMutation,
  AddS3ObjectToCollectionMutationVariables,
} from "./collections.generated";

export async function addToCollection(
  collectionId: string,
  element: { id: string; __typename: string }
) {
  const client = getApolloClient();
  switch (element.__typename) {
    case "DHIS2DataElement": {
      const { data } = await client.mutate<
        AddDhis2DataElementToCollectionMutation,
        AddDhis2DataElementToCollectionMutationVariables
      >({
        mutation: gql`
          mutation addDHIS2DataElementToCollection(
            $input: AddDHIS2DataElementToCollectionInput!
          ) {
            addToCollection: addDHIS2DataElementToCollection(input: $input) {
              success
              errors
            }
          }
        `,
        variables: { input: { id: element.id, collectionId } },
      });
      return data?.addToCollection.success;
    }
    case "S3Object": {
      const { data } = await client.mutate<
        AddS3ObjectToCollectionMutation,
        AddS3ObjectToCollectionMutationVariables
      >({
        mutation: gql`
          mutation addS3ObjectToCollection(
            $input: AddS3ObjectToCollectionInput!
          ) {
            addToCollection: addS3ObjectToCollection(input: $input) {
              success
              errors
            }
          }
        `,
        variables: { input: { id: element.id, collectionId } },
      });
      return data?.addToCollection.success;
    }
  }
}

export async function removeFromCollection(
  collectionId: string,
  element: { id: string; __typename: string }
) {
  const client = getApolloClient();
  switch (element.__typename) {
    case "DHIS2DataElement": {
      const { data } = await client.mutate({
        mutation: gql`
          mutation removeDHIS2DataElementFromCollection(
            $input: RemoveDHIS2DataElementFromCollectionInput!
          ) {
            removeFromCollection: removeDHIS2DataElementFromCollection(
              input: $input
            ) {
              success
              errors
            }
          }
        `,
        variables: {
          input: {
            collectionId,
            id: element.id,
          },
        },
      });
      return data?.removeFromCollection.success;
    }
    case "S3Object": {
      const { data } = await client.mutate({
        mutation: gql`
          mutation removeS3ObjectFromCollection(
            $input: RemoveS3ObjectFromCollectionInput!
          ) {
            removeFromCollection: removeS3ObjectFromCollection(input: $input) {
              success
              errors
            }
          }
        `,
        variables: {
          input: {
            collectionId,
            id: element.id,
          },
        },
      });
      return data?.removeFromCollection.success;
    }
  }
}

export async function createCollection(input: CreateCollectionInput) {
  const client = getApolloClient();
  const { data } = await client.mutate<
    CreateCollectionMutation,
    CreateCollectionMutationVariables
  >({
    mutation: gql`
      mutation CreateCollection($input: CreateCollectionInput!) {
        createCollection(input: $input) {
          success
          errors
          collection {
            id
            name
          }
        }
      }
    `,
    variables: { input },
  });

  if (!data?.createCollection.collection) {
    throw new Error("Unable to create collection.");
  }

  return data?.createCollection.collection;
}
