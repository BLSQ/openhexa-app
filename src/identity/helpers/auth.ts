import { getApolloClient } from "core/helpers/apollo";
import {
  DisableTwoFactorDocument,
  DisableTwoFactorMutation,
  DisableTwoFactorMutationVariables,
  EnableTwoFactorDocument,
  EnableTwoFactorMutation,
  GenerateChallengeDocument,
  GenerateChallengeMutation,
  VerifyDeviceDocument,
  VerifyDeviceMutation,
  VerifyDeviceMutationVariables,
} from "identity/graphql/mutations.generated";
import {
  GetUserDocument,
  GetUserQuery,
} from "identity/graphql/queries.generated";
import { GetServerSidePropsContext } from "next";
import { getCookie } from "cookies-next";
import Router from "next/router";

export async function getMe(ctx?: GetServerSidePropsContext) {
  const client = getApolloClient({ headers: ctx?.req.headers });
  const payload = await client.query<GetUserQuery>({
    query: GetUserDocument,
  });
  const me = payload?.data.me ?? {};
  if (!me) return null;
  return { ...me };
}

export async function logout() {
  let csrfToken = getCookie("csrftoken");
  if (typeof csrfToken !== "string") {
    throw new Error("Could not find CSRF token in cookies.");
  }
  fetch("/auth/logout/", {
    method: "POST",
    redirect: "follow",
    headers: { "X-CSRFToken": csrfToken },
  }).then((response) => {
    return Router.replace(response.url);
  });
}

export async function generateChallenge(ctx?: GetServerSidePropsContext) {
  const client = getApolloClient({ headers: ctx?.req.headers });

  const payload = await client.mutate<GenerateChallengeMutation>({
    mutation: GenerateChallengeDocument,
  });

  return payload.data?.generateChallenge.success ?? false;
}

export async function disableTwoFactor(token: string) {
  const client = getApolloClient();

  const payload = await client.mutate<
    DisableTwoFactorMutation,
    DisableTwoFactorMutationVariables
  >({
    mutation: DisableTwoFactorDocument,
    variables: { input: { token } },
  });

  return payload.data?.disableTwoFactor.success ?? false;
}

export async function enableTwoFactor() {
  const client = getApolloClient();

  const payload = await client.mutate<EnableTwoFactorMutation>({
    mutation: EnableTwoFactorDocument,
  });

  if (!payload.data) {
    throw new Error("Unable to enable two factor authentication");
  }

  return payload.data.enableTwoFactor;
}

export async function verifyDevice(token?: string) {
  const client = getApolloClient();

  const payload = await client.mutate<
    VerifyDeviceMutation,
    VerifyDeviceMutationVariables
  >({
    mutation: VerifyDeviceDocument,
    variables: { input: { token } },
  });

  if (!payload.data) {
    throw new Error("Unable to enable two factor authentication");
  }

  return payload.data.verifyDevice.success;
}
