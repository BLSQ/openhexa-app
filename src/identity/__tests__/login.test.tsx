import { MockedResponse } from "@apollo/client/testing";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import {
  LoginDocument,
  useLoginMutation,
} from "identity/graphql/mutations.generated";
import { TestApp } from "core/helpers/testutils";
import mockRouter from "next-router-mock";
import router from "next/router";
import LoginPage from "pages/login";

jest.mock("identity/graphql/mutations.generated", () => {
  const actualModule = jest.requireActual(
    "identity/graphql/mutations.generated",
  );
  return {
    ...actualModule,
    __esModule: true,
    useLoginMutation: jest.fn(actualModule.useLoginMutation),
  };
});

const useLoginMutationMock = useLoginMutation as jest.Mock;
describe("LoginPage: No Two Factor Authentication", () => {
  beforeEach(() => {
    mockRouter.setCurrentUrl("/");
    useLoginMutationMock.mockClear();
  });

  it("renders the login page", async () => {
    const { container } = render(
      <TestApp>
        <LoginPage />
      </TestApp>,
    );

    expect(container).toMatchSnapshot();
  });

  it("calls the login action on submit", async () => {
    const doLogin = jest.fn();
    useLoginMutationMock.mockReturnValue([doLogin]);
    const user = userEvent.setup();
    const { container } = render(
      <TestApp>
        <LoginPage />
      </TestApp>,
    );

    const emailInput = screen.getByTestId("email");
    const passwordInput = screen.getByTestId("password");
    const submitBtn = screen.getByRole("button", { name: "Sign in" });

    await user.type(emailInput, "root@openhexa.org");
    await user.click(submitBtn);
    expect(doLogin).not.toHaveBeenCalled();

    await user.type(passwordInput, "pA$$W0rd");
    await user.click(submitBtn);
    expect(useLoginMutation).toHaveBeenCalled();
    expect(doLogin).toHaveBeenCalledWith({
      variables: {
        input: {
          email: "root@openhexa.org",
          password: "pA$$W0rd",
          token: undefined,
        },
      },
    });
  });

  it("redirects the user on success", async () => {
    const pushSpy = jest.spyOn(router, "push");
    const { useLoginMutation } = jest.requireActual(
      "identity/graphql/mutations.generated",
    );
    useLoginMutationMock.mockImplementation(useLoginMutation);
    const user = userEvent.setup();
    const mocks = [
      {
        request: {
          query: LoginDocument,
          variables: {
            input: {
              email: "root@openhexa.org",
            },
          },
        },
        result: {
          data: { login: { success: true } },
        },
      },
    ];
    expect(useLoginMutationMock).not.toHaveBeenCalled();
    render(
      <TestApp mocks={mocks}>
        <LoginPage />
      </TestApp>,
    );
    expect(useLoginMutationMock).toHaveBeenCalled();

    const emailInput = screen.getByTestId("email");
    const passwordInput = screen.getByTestId("password");
    const submitBtn = screen.getByRole("button", { name: "Sign in" });

    await user.type(emailInput, "root@openhexa.org");
    await user.click(submitBtn);
    expect(pushSpy).not.toHaveBeenCalled();
    await user.type(passwordInput, "pA$$W0rd");
    await user.click(submitBtn);
    waitFor(() => {
      expect(pushSpy).toHaveBeenCalled();
    });
  });

  it("displays an error message if email/password is incorrect", async () => {
    const { useLoginMutation } = jest.requireActual(
      "identity/graphql/mutations.generated",
    );
    useLoginMutationMock.mockImplementation(useLoginMutation);
    const user = userEvent.setup();
    const [EMAIL, PWD] = ["root@openhexa.org", "pA$$W0rd"];
    const mocks = [
      {
        request: {
          query: LoginDocument,
          variables: {
            input: {
              email: EMAIL,
              password: PWD,
            },
          },
        },
        result: {
          data: { login: { success: false, errors: ["INVALID_CREDENTIALS"] } },
        },
      },
    ];
    render(
      <TestApp mocks={mocks}>
        <LoginPage />
      </TestApp>,
    );

    const emailInput = screen.getByTestId("email");
    const passwordInput = screen.getByTestId("password");
    const submitBtn = screen.getByRole("button", { name: "Sign in" });
    await user.type(emailInput, EMAIL);
    await user.type(passwordInput, PWD);
    await user.click(submitBtn);
    const errorMessage = await screen.findByText(
      "Wrong email address and/or password.",
    );

    expect(errorMessage).toBeInTheDocument();
  });
});

describe("LoginPage: With Two Factor Authentication", () => {
  beforeEach(() => {
    mockRouter.setCurrentUrl("/");
    useLoginMutationMock.mockImplementation(useLoginMutation);
  });

  it("asks for a token after a login attempt", async () => {
    const doLogin = jest.fn(() => ({
      data: {
        login: { success: false, errors: ["OTP_REQUIRED"] },
      },
    }));
    useLoginMutationMock.mockReturnValue([doLogin]);
    expect(doLogin).not.toHaveBeenCalled();
    const user = userEvent.setup();
    const [EMAIL, PWD] = ["root@openhexa.org", "pA$$W0rd"];

    render(
      <TestApp me={{ hasTwoFactorEnabled: true }}>
        <LoginPage />
      </TestApp>,
    );

    const emailInput = screen.getByTestId("email");
    const passwordInput = screen.getByTestId("password");
    const submitBtn = screen.getByRole("button", { name: "Sign in" });
    await user.type(emailInput, EMAIL);
    await user.type(passwordInput, PWD);
    await user.click(submitBtn);
    expect(doLogin).toHaveBeenCalled();

    const message = await screen.findByText(
      "Enter the OTP code you received in your mailbox.",
    );
    expect(message).toBeInTheDocument();

    const tokenInput = screen.getByTestId("token");
    expect(tokenInput).toBeInTheDocument();
    await user.type(tokenInput, "121212");
    await user.click(submitBtn);

    expect(doLogin).toHaveBeenLastCalledWith({
      variables: {
        input: {
          email: EMAIL,
          password: PWD,
          token: "121212",
        },
      },
    });
  });
});
