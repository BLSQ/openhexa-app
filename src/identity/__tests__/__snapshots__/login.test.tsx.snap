// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage: No Two Factor Authentication renders the login page 1`] = `
<div>
  <form
    class="max-w-md flex-1 space-y-6"
  >
    <div>
      <div
        class="relative h-16 w-auto"
      >
        <span
          style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
        >
          <img
            alt="OpenHEXA logo"
            class="mx-auto block h-16 w-auto"
            data-nimg="fill"
            decoding="async"
            sizes="100vw"
            src="/images/logo.svg"
            srcset="/images/logo.svg 640w, /images/logo.svg 750w, /images/logo.svg 828w, /images/logo.svg 1080w, /images/logo.svg 1200w, /images/logo.svg 1920w, /images/logo.svg 2048w, /images/logo.svg 3840w"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
        </span>
      </div>
      <h2
        class="mt-6 text-center text-3xl font-extrabold text-gray-900"
      >
        Sign in
      </h2>
    </div>
    <div
      class="-space-y-px pt-2"
    >
      <label
        class="sr-only"
        for="email"
      >
        Email address
      </label>
      <div
        class="group relative w-full"
      >
        <input
          aria-describedby="email-description"
          aria-invalid="false"
          autocomplete="email"
          class="form-input appearance-none relative block px-3 py-2 border rounded-md focus:outline-hidden focus:z-10 sm:text-sm disabled:opacity-50 disabled:cursor-not-allowed border-gray-300 placeholder-gray-500 text-gray-900 focus:ring-blue-500 focus:border-blue-500 w-full rounded-b-none"
          data-testid="email"
          id="email"
          name="email"
          placeholder="Email address"
          required=""
          type="text"
          value=""
        />
      </div>
      <label
        class="sr-only"
        for="password"
      >
        Password
      </label>
      <div
        class="group relative w-full"
      >
        <input
          aria-describedby="password-description"
          aria-invalid="false"
          autocomplete="current-password"
          class="form-input appearance-none relative block px-3 py-2 border rounded-md focus:outline-hidden focus:z-10 sm:text-sm disabled:opacity-50 disabled:cursor-not-allowed border-gray-300 placeholder-gray-500 text-gray-900 focus:ring-blue-500 focus:border-blue-500 w-full rounded-t-none"
          data-testid="password"
          id="password"
          name="password"
          placeholder="Password"
          required=""
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      class="flex items-center justify-end"
    >
      <div
        class="text-sm"
      >
        <a
          class="text-blue-600 hover:text-blue-500"
          href="/auth/password_reset"
          rel="noopener noreferrer"
        >
          Forgot your password?
        </a>
      </div>
    </div>
    <div
      class="space-y-2"
    >
      <button
        class="relative h-fit
    inline-flex items-center justify-center transition-all
    border font-medium
    disabled:opacity-50 disabled:cursor-not-allowed
   rounded-sm focus:outline-hidden focus:ring-2 focus:ring-offset-2 text-sm px-3 py-2.5 leading-4 shadow-xs border-transparent text-white bg-blue-600 hover:bg-blue-700 focus:ring-blue-500 w-full"
        data-testid="submit"
        type="submit"
      >
        Sign in
      </button>
    </div>
  </form>
</div>
`;
